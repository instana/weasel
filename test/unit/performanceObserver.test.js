import sinon from 'sinon';

jest.mock('../../lib/browser');
jest.mock('../../lib/performance');

describe.only('performanceObserver', () => {
  let browserMock;
  let performanceMock;
  let observeResourcePerformance;
  let PerformanceObserverConstructorArgs;
  let PerformanceObserverMock;
  let resourceMatcher;
  let onEnd;
  let clock;

  beforeEach(() => {
    clock = sinon.useFakeTimers();
    resourceMatcher = jest.fn();
    resourceMatcher.mockReturnValue(true);
    browserMock = require('../../lib/browser');
    browserMock.reset();
    PerformanceObserverMock = {
      observe: jest.fn(),
      disconnect: jest.fn()
    };
    PerformanceObserverConstructorArgs = jest.fn();
    browserMock.win.PerformanceObserver = function(opts) {
      PerformanceObserverConstructorArgs(opts);
      return PerformanceObserverMock;
    };

    performanceMock = require('../../lib/performance');
    performanceMock.reset();
    observeResourcePerformance = require('../../lib/performanceObserver').observeResourcePerformance;
    onEnd = jest.fn();
  });

  afterEach(() => {
    clock.restore();
  });

  it('must work when performance observer is not available', () => {
    performanceMock.setPerformanceObserverAvailable(false);
    const observer = observeResourcePerformance({
      entryTypes: 'resource',
      resourceMatcher:  () => true,
      maxWaitForResourceMillis: 1000,
      onEnd
    });

    observer.onBeforeResourceRetrieval();
    clock.tick(42);
    observer.onAfterResourceRetrieved();

    expect(onEnd).toMatchSnapshot();
    expect(PerformanceObserverConstructorArgs).toMatchSnapshot();
  });

  it('must handle performance timing before resource completes', () => {
    const observer = observeResourcePerformance({
      entryTypes: 'resource',
      resourceMatcher,
      maxWaitForResourceMillis: 1000,
      onEnd
    });

    observer.onBeforeResourceRetrieval();
    simulateResource({
      name: 'customResource',
      startTime: Date.now(),
      responseEnd: Date.now()
    });
    observer.onAfterResourceRetrieved();

    expect(onEnd).toMatchSnapshot();
    expect(browserMock.win.PerformanceObserver).toMatchSnapshot();
  });

  function simulateResource(resource) {
    expect(PerformanceObserverConstructorArgs.mock.calls.length).toEqual(1);
    PerformanceObserverConstructorArgs.mock.calls[0][0]({
      getEntries() {
        return [resource];
      }
    });
  }
});
