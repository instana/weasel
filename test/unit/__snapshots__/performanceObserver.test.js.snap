// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performanceObserver must handle performance timing before resource completes 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "duration": 0,
        "resource": Object {
          "name": "customResource",
          "responseEnd": 0,
          "startTime": 0,
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`performanceObserver must handle performance timing before resource completes 2`] = `[Function]`;

exports[`performanceObserver must work when performance observer is not available 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "duration": 42,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`performanceObserver must work when performance observer is not available 2`] = `[MockFunction]`;
